consensus/hotstuff/vote_collector.go

```diff
@@ -0,0 +1,92 @@
+var collectorStatusNames = [...]string{"VoteCollectorStatusCaching",
+	"VoteCollectorStatusVerifying",
+	"VoteCollectorStatusInvalid"}
+
+func (ps VoteCollectorStatus) String() string {
+	if ps < 0 || int(ps) > len(collectorStatusNames) {
+		return "UNKNOWN"
+	}
+	return collectorStatusNames[ps]
+}
+
+type VoteCollector interface {
+	VoteCollectorState
+	// ChangeProcessingStatus changes the VoteCollector's internal processing
+	// status. The operation is implemented as an atomic compare-and-swap, i.e. the
+	// state transition is only executed if VoteCollector's internal state is
+	// equal to `expectedValue`. The return indicates whether the state was updated.
+	// The implementation only allows the transitions
+	//         CachingVotes   -> VerifyingVotes
+	//         CachingVotes   -> Invalid
+	//         VerifyingVotes -> Invalid
+	// Error returns:
+	// * nil if the state transition was successfully executed
+	// * ErrDifferentCollectorState if the VoteCollector's state is different than expectedCurrentStatus
+	// * ErrInvalidCollectorStateTransition if the given state transition is impossible
+	// * all other errors are unexpected and potential symptoms of internal bugs or state corruption (fatal)
+	ChangeProcessingStatus(expectedValue, newValue VoteCollectorStatus) error
+
+	// ProcessBlock performs validation of block signature and processes block with respected collector.
+	// Calling this function will mark conflicting collector as stale and change state of valid collectors
+	// It returns nil if the block is valid.
+	// It returns model.InvalidBlockError if block is invalid.
+	// It returns other error if there is exception processing the block.
+	ProcessBlock(block *model.Proposal) error
+}
+
+// VoteCollectorState collects votes for the same block, produces QC when enough votes are collected
+// VoteCollectorState takes a callback function to report the event that a QC has been produced.
+type VoteCollectorState interface {
+	// AddVote adds a vote to the collector
+	// return error if the signature is invalid
+	// When enough votes have been added to produce a QC, the QC will be created asynchronously, and
+	// passed to EventLoop through a callback.
+	AddVote(vote *model.Vote) error
+
+	// BlockID returns the block ID that this instance is collecting votes for.
+	// This method is useful when adding the newly created vote collector to vote collectors map.
+	BlockID() flow.Identifier
+
+	// Status returns the status of the vote collector
+	Status() VoteCollectorStatus
+}
+
+// VerifyingVoteCollector is a VoteCollector and also implement the same interface as BlockSigner, so that
+// when the voter ask VoteAggregator(via BlockSigner interface)
+// to sign the block, and VoteAggregator will read the VerifyingVoteCollector from the vote collectors
+// map and produce the vote.
+// Note CachingVoteCollector can't create vote, only VerifyingVoteCollector can
+type VerifyingVoteCollector interface {
```

Comment:
Not sure we need to expose `VerifyingVoteCollector` at this level

Reply:
The StateMachine will see this interface instead of a concrete struct, such that the collection cluster could provide a different implementation